src/service/dashboards.rs

```diff
@@ -12,78 +12,107 @@
-        Err(_) => Ok(HttpResponse::NotFound().json(MetaHttpResponse::error(
-            StatusCode::NOT_FOUND.into(),
-            "Dashboard not found".to_string(),
-        ))),
+impl From<Response> for HttpResponse {
+    fn from(resp: Response) -> Self {
+        match resp {
+            Response::Created { dashboard_id } => Self::Created().json(MetaHttpResponse::message(
+                StatusCode::CREATED.into(),
+                format!("Dashboard {dashboard_id} created"),
+            )),
+            Response::NotFound { error } => Self::NotFound().json(MetaHttpResponse::error(
+                StatusCode::NOT_FOUND.into(),
+                error.unwrap_or("Dashboard not found".to_owned()),
+            )),
+            Response::InternalServerError(err) => Self::InternalServerError().json(
+                MetaHttpResponse::error(StatusCode::INTERNAL_SERVER_ERROR.into(), err.to_string()),
+            ),
+        }
     }
 }
 
-pub async fn save_dashboard(
+#[instrument(skip(dashboard))]
+pub async fn create_dashboard(
     org_id: &str,
-    name: &str,
-    dashboard: &Dashboard,
+    dashboard: Dashboard,
 ) -> Result<HttpResponse, io::Error> {
-    let loc_span = info_span!("service:dashboards:save");
-    let _guard = loc_span.enter();
+    // XXX-TODO(vvv): Generate[*] `dashboard_id` and overwrite
+    // `dashboard.dashboard_id` with this value.
+    //
+    // [*]: Use global counter, stored in the database. Increment the counter
+    // on each `create_dashboard` call.
+    let resp = if let Err(e) = dashboard::put(org_id, &dashboard).await {
+        Response::InternalServerError(e)
+    } else {
+        Response::Created {
+            dashboard_id: dashboard.dashboard_id,
+        }
+    };
+    Ok(resp.into())
+}
 
-    match dashboard::set(org_id, name, dashboard).await {
-        Ok(_) => Ok(HttpResponse::Ok().json(MetaHttpResponse::message(
-            http::StatusCode::OK.into(),
-            "Dashboard saved".to_string(),
-        ))),
-        Err(e) => Ok(
-            HttpResponse::InternalServerError().json(meta::http::HttpResponse::error(
-                StatusCode::BAD_REQUEST.into(),
-                e.to_string(),
-            )),
-        ),
+#[instrument(skip(dashboard))]
+pub async fn update_dashboard(
+    org_id: &str,
+    dashboard_id: &str,
+    dashboard: &Dashboard,
+) -> Result<HttpResponse, io::Error> {
+    if dashboard::get(org_id, dashboard_id).await.is_err() {
+        return Ok(Response::NotFound { error: None }.into());
     }
+    if let Err(e) = dashboard::put(org_id, dashboard).await {
+        return Ok(Response::InternalServerError(e).into());
+    }
+    Ok(HttpResponse::Ok().json(MetaHttpResponse::message(
+        StatusCode::OK.into(),
+        "Dashboard updated".to_owned(),
+    )))
 }
 
+#[instrument]
 pub async fn list_dashboards(org_id: &str) -> Result<HttpResponse, io::Error> {
-    let loc_span = info_span!("service:dashboards:list");
-    let _guard = loc_span.enter();
```

Comment:
See https://docs.rs/tracing/0.1.37/tracing/span/struct.Span.html#in-asynchronous-code

Reply:
Explicit naming of spans comes with a higher risk of using the same name (label) for different functions due to a copy-paste mistake. Compiler won't prevent you from writing

```rust
#[instrument(name = "foo")]
fn f() {
    // ...
}

#[instrument(name = "foo")]
fn g() {
    // ...
}
```

Currently the log lines look like this:
```
[2023-04-05T06:45:24Z INFO  zincobserve::service::dashboards] list_dashboards; org_id="default"
[2023-04-05T06:45:24Z INFO  zincobserve::service::db::dashboard] list; org_id="default"
```

@hengfeiyang Are there any issues with the auto-naming of spans that I'm missing?